mily/runengine.py

```diff
@@ -0,0 +1,305 @@
+
+    Listens to the state of the RunEngine and shows the available commands for
+    the current state.
+
+    Attributes
+    ----------
+
+    available_commands: dict
+        Mapping of state to available RunEngine commands
+    """
+    available_commands = {'running': ['Halt', 'Pause'],
+                          'idle': ['Start'],
+                          'paused': ['Abort', 'Halt',  'Resume', 'Stop']}
+
+    command_flags = {'Halt': QDialogButtonBox.DestructiveRole,
+                     'Pause': QDialogButtonBox.HelpRole,
+                     'Start': QDialogButtonBox.AcceptRole,
+                     'Abort': QDialogButtonBox.RejectRole,
+                     'Resume': QDialogButtonBox.AcceptRole,
+                     'Stop': QDialogButtonBox.ResetRole}
+
+    def __init__(self, *args, **kwargs):
+        super(EngineControl, self).__init__(*args, **kwargs)
+        self.buttonGroup = QButtonGroup()
+        self._buttons = {name: QPushButton(name) for command in self.available_commands.values() for name in command}
+        for name, button in self._buttons.items():
+            self.addButton(button, self.command_flags[name])
+            self.buttonGroup.addButton(button)
+
+    @Slot('QString', 'QString')
+    def on_state_change(self, state, old_state):
+        for name, button in self._buttons.items():
+            button.setEnabled(name in self.available_commands[state])
+
+
+class EngineWidget(QWidget):
+    """
+    RunEngine Control Widget
+
+    Parameters
+    ----------
+    engine : RunEngine, optional
+        The underlying RunEngine object. A basic version wil be instatiated if
+        one is not provided
+
+    plan : callable, optional
+        A callable  that takes no parameters and returns a generator. If the
+        plan is meant to be called repeatedly the function should make sure
+        that a refreshed generator is returned each time
+
+    Attributes
+    ----------
+    engine_state_change : Signal('QString', 'QString')
+        Signal emitted by changes in the RunEngine state. The first string is
+        the current state, the second is the previous state
+
+    update_rate: float
+        Update rate the qt_kicker is installed at
+
+    command_registry: dict
+        Mapping of commands received by the Slot `command` and actual
+        Python callables
+    """
+    engine_state_change = Signal('QString', 'QString')
+    update_rate = 0.02
+    command_registry = {'Halt': RunEngine.halt,
+                        'Start': no_plan_warning,
+                        'Abort': RunEngine.abort,
+                        'Resume': RunEngine.resume,
+                        'Pause': RunEngine.request_pause,
+                        'Stop': RunEngine.stop}
+
+    def __init__(self, engine=None, plan=None, parent=None):
+        # Instantiate widget information and layout
+        super().__init__(parent=parent)
+
+        self.setStyleSheet('QLabel {qproperty-alignment: AlignCenter}')
```

Comment:
I wonder if mily should make a consistent set of rules as to how `QWidget` subclasses handle style sheets.

Reply:
Harmless, sure - but not so easily extensible.

I doubt it's a controversial opinion, but I think everything in mily should be designed with extensibility in mind, where possible.